sympy/core/symbol.py

```diff
@@ -737,3 +739,47 @@ def traverse(symbols, frame):
         del frame  # break cyclic dependencies as stated in inspect docs
 
     return syms
+
+def disambiguate(*iter):
+    """
+    This provides a method to disambiguate symbols or expressions having
+    ambiguate symbols
+
+    Parameters
+    ==========
+
+    iter: list of symbols or expressions.
+
+    Returns
+    =======
+
+    An ordered tuple containing disambiguate symbols or expressions
+    with disambiguate symbols.
+
+    Examples
+    ========
+
+    >>> from sympy.core.symbol import disambiguate
+    >>> from sympy import Dummy, Symbol
+
+    >>> tup = Symbol('_x'), Dummy('x'), Dummy('x')
+    >>> disambiguate(*tup)
+    (_x, _x_1, _x_2)
+
+    >>> disambiguate(Dummy('x'), Dummy('x'))
+    (_x, _x_1)
+
+    """
+    new_iter = Tuple(*iter)
+    syms = ordered(new_iter.free_symbols)
+    mapping = {}
+    for s in syms:
+        mapping.setdefault(str(s), []).append(s)
+    reps = {}
+    for k in mapping:
+        reps[mapping[k][0]] = Symbol("%s"%(k), **mapping[k][0].assumptions0)
+        if len(mapping[k]) == 1:
+            continue
+        for i in range(1, len(mapping[k])):
+            reps[mapping[k][i]] = Symbol("%s_%i"%(k, i), **mapping[k][i].assumptions0)
+    return tuple(ordered(new_iter.xreplace(reps)))
```

Comment:
OK, I see you are doing the same as I intended...I got thrown off by the examples. And btw, I just ran into an expression -- though can't reproduce from memory -- where a minpoly had two "x" values in the result, but they we actually different variables.

A fewcomments:
1. if you don't sort then the numbering will be arbitrary and not canonical; the demo of the `ok` function shows an example wherein the (arbitrary) ordering of the symbols doesn't lead to a canonical result.
2. you definitely don't want to sort the result coming back -- someone will want to see expr[0] still in slot 0 with its new symbols; I suspect that's why you thought ordering symbols (point previous) was not necessary.
3. perhaps you could make the mapping (potentially inverted) be an optional argument; with it, you can replace the symbols with whatever you want once the routine has figured out which symbols might need to be tweaked. How could it be made a defaultdict that gives back the key if a value for that key is missing (other than doing the obvious as a user: `result = mapping.get(x, x)`)...[googling](https://stackoverflow.com/questions/10524142/python-dictionary-return-requested-key-if-value-does-not-exist):
```
from collections import defaultdict

class KeyAwareDefaultDict(defaultdict):
    def __missing__(self, key):
        if self.default_factory is None:
            raise KeyError(key)
        self[key] = value = self.default_factory(key)
        return value

>>> lookup = KeyAwareDefaultDict(Id)
>>> lookup[x]
x
>>> lookup[x] = y; lookup[x]
y
```

Reply:
For point 1: Yes to make it canonical, I have `ordered` the free_symbols at the start.

For point 2: Actually, an ordered tuple is returned as the answer, and this changes the ordering as it was in the input. I just did to make the output more simpler to read. If you suggest I could remove ordered in the last return statement.

For point 3: We can use as you suggest, but I am not sure whether it would make any difference. Means, it would reduce the loops but will it affect any performance of the function?, otherwise it is just another way to implement the function.(*I will try to implement it locally and check)